## Chapter 6

In [1]:
import pandas as pd
import statsmodels.api as sm
import numpy as np

In [2]:
#Exercise C1
kielmc = pd.read_stata("stata/KIELMC.DTA")
kielmc_1981 = kielmc[kielmc.year == 1981]

y = kielmc_1981.lprice
X = sm.add_constant(kielmc_1981[["ldist"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 lprice   R-squared:                       0.180
Model:                            OLS   Adj. R-squared:                  0.174
Method:                 Least Squares   F-statistic:                     30.79
Date:                Sun, 24 May 2020   Prob (F-statistic):           1.40e-07
Time:                        23:28:04   Log-Likelihood:                -53.138
No. Observations:                 142   AIC:                             110.3
Df Residuals:                     140   BIC:                             116.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          8.0472      0.646     12.452      0.0

In [3]:
X = sm.add_constant(kielmc_1981[["ldist", "lintst", "larea", "lland", "rooms", "baths", "age"]])

model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 lprice   R-squared:                       0.748
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     56.68
Date:                Sun, 24 May 2020   Prob (F-statistic):           5.10e-37
Time:                        23:28:04   Log-Likelihood:                 30.482
No. Observations:                 142   AIC:                            -44.96
Df Residuals:                     134   BIC:                            -21.32
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.5923      0.642     11.831      0.0

In [4]:
kielmc_1981["lintst2"] = kielmc_1981.lintst ** 2
X = sm.add_constant(kielmc_1981[["ldist", "lintst", "larea", "lland", "rooms", "baths", "age", "lintst2"]])

model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 lprice   R-squared:                       0.778
Model:                            OLS   Adj. R-squared:                  0.764
Method:                 Least Squares   F-statistic:                     58.11
Date:                Sun, 24 May 2020   Prob (F-statistic):           9.95e-40
Time:                        23:28:04   Log-Likelihood:                 39.466
No. Observations:                 142   AIC:                            -60.93
Df Residuals:                     133   BIC:                            -34.33
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -3.3171      2.646     -1.254      0.2

/home/systemchalk/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [5]:
kielmc_1981["ldist2"] = kielmc_1981.ldist ** 2
X = sm.add_constant(kielmc_1981[["ldist", "lintst", "larea", "lland", "rooms", "baths", "age", "lintst2", "ldist2"]])

model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 lprice   R-squared:                       0.778
Model:                            OLS   Adj. R-squared:                  0.763
Method:                 Least Squares   F-statistic:                     51.32
Date:                Sun, 24 May 2020   Prob (F-statistic):           7.44e-39
Time:                        23:28:04   Log-Likelihood:                 39.525
No. Observations:                 142   AIC:                            -59.05
Df Residuals:                     132   BIC:                            -29.49
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -5.9213      8.295     -0.714      0.4

/home/systemchalk/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


C1.i This suggests prices increase with distance from the incenerator (implying that the incenerator decreases values)

C1.ii The effects of the incinerator now vanish due to other factors associated with house prices being accounted for

C1.iii ldist becomes significant. Functional form matters!

C1.iv It is not

In [6]:
#Exercise C2
wage1 = pd.read_stata("stata/WAGE1.DTA")

y = wage1.lwage
X = sm.add_constant(wage1[["educ", "exper", "expersq"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.300
Model:                            OLS   Adj. R-squared:                  0.296
Method:                 Least Squares   F-statistic:                     74.67
Date:                Sun, 24 May 2020   Prob (F-statistic):           3.38e-40
Time:                        23:28:05   Log-Likelihood:                -319.53
No. Observations:                 526   AIC:                             647.1
Df Residuals:                     522   BIC:                             664.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.1280      0.106      1.208      0.2

In [7]:
print(100 * (model.params[2] + (2 * (model.params[3] * 5))))
print(100 * (model.params[2] + (2 * (model.params[3] * 20))))

3.3873293897328556
1.2466549129434912


In [8]:
abs(model.params[2] / (2 * model.params[3]))

28.735482156165478

In [9]:
wage1[wage1.exper > 28.8].shape[0]

121

C2.ii It is significant at the 1% level

C2.iii 3.39%; 1.25%

C2.iv 28.74; 121 observations have a higher level of experience than this

In [10]:
#Exercise C3
wage2 = pd.read_stata("stata/WAGE2.DTA")
wage2["educexper"] = wage2.educ.astype("int") * wage2.exper.astype("int")

y = wage2.lwage
X = sm.add_constant(wage2[["educ", "exper", "educexper"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.135
Model:                            OLS   Adj. R-squared:                  0.132
Method:                 Least Squares   F-statistic:                     48.41
Date:                Sun, 24 May 2020   Prob (F-statistic):           4.47e-29
Time:                        23:28:05   Log-Likelihood:                -449.87
No. Observations:                 935   AIC:                             907.7
Df Residuals:                     931   BIC:                             927.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.9495      0.241     24.704      0.0

In [11]:
wage2["b3_var"] = wage2.educexper.astype("int") - (10 * wage2.educ.astype("int"))

y = wage2.lwage
X = sm.add_constant(wage2[["educ", "exper", "b3_var"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.135
Model:                            OLS   Adj. R-squared:                  0.132
Method:                 Least Squares   F-statistic:                     48.41
Date:                Sun, 24 May 2020   Prob (F-statistic):           4.47e-29
Time:                        23:28:05   Log-Likelihood:                -449.87
No. Observations:                 935   AIC:                             907.7
Df Residuals:                     931   BIC:                             927.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.9495      0.241     24.704      0.0

C3.i Using calculus, $\frac{d}{deduc} = \beta_1 + \beta_3 exper$

C3.ii $\beta_2 = 0$ alternative is $\beta_2 \neq 0$

C3.iii We reject the null

C3.iv 0.0761, [0.063, 0.089]

In [12]:
#Exercise C4
gpa2 = pd.read_excel("excel/gpa2.xls", 
                     names = ["sat", "tothrs", "colgpa", "athlete", "verbmath", "hsize", "hsrank", "hsperc", "female", "white", "black", "hsizesq"],
                     sep = " ")

y = gpa2.sat
X = sm.add_constant(gpa2[["hsize", "hsizesq"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                    sat   R-squared:                       0.008
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     15.78
Date:                Sun, 24 May 2020   Prob (F-statistic):           1.49e-07
Time:                        23:28:05   Log-Likelihood:                -26274.
No. Observations:                4136   AIC:                         5.255e+04
Df Residuals:                    4133   BIC:                         5.257e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        998.1312      6.209    160.744      0.0

In [13]:
model.params[1] / (2 * abs(model.params[2]))

4.651379609491661

In [14]:
gpa2["lsat"] = np.log(gpa2.sat)
y = gpa2.lsat

model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                   lsat   R-squared:                       0.008
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     16.05
Date:                Sun, 24 May 2020   Prob (F-statistic):           1.14e-07
Time:                        23:28:05   Log-Likelihood:                 2331.6
No. Observations:                4136   AIC:                            -4657.
Df Residuals:                    4133   BIC:                            -4638.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.8962      0.006   1119.962      0.0

In [15]:
model.params[1] / (2 * abs(model.params[2]))

4.69745656429681

C4.i The quadratic term is statistically significant

C4.ii The optimal high school size is 465 as the quadratic term overwhelms the positive term beyond this point

C4.iii It is not clear where this GPA data comes from but 4136 observations without any information as to districts or states it is difficult to see how this is a convincing general statement

C4.iv The optimal class size in logs is 469, not substantially different from the level regression

In [16]:
#Exercise C5
hprice1 = pd.read_stata("stata/hprice1.dta")

y = hprice1.lprice
X = sm.add_constant(hprice1[["llotsize", "lsqrft", "bdrms"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 lprice   R-squared:                       0.643
Model:                            OLS   Adj. R-squared:                  0.630
Method:                 Least Squares   F-statistic:                     50.42
Date:                Sun, 24 May 2020   Prob (F-statistic):           9.74e-19
Time:                        23:28:06   Log-Likelihood:                 25.861
No. Observations:                  88   AIC:                            -43.72
Df Residuals:                      84   BIC:                            -33.81
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.2970      0.651     -1.992      0.0

In [17]:
X_p = pd.DataFrame({"const" : [1.0], "lotsize" : [np.log(20000)], "lsqrft": [np.log(2500)], "bdrms": [4]})
print(model.predict(exog=X_p))

0    5.992898
dtype: float64


In [18]:
(sum(np.exp(model.resid)) / model.nobs) * np.exp(model.predict(exog=X_p))

0    407.122394
dtype: float64

In [19]:
((np.exp(model.fittedvalues) * np.exp(hprice1.lprice)).sum() / (np.exp(model.fittedvalues)**2).sum()) * np.exp(model.predict(exog=X_p))

0    409.752133
dtype: float64

In [20]:
y = hprice1.price
X = sm.add_constant(hprice1[["lotsize", "sqrft", "bdrms"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.672
Model:                            OLS   Adj. R-squared:                  0.661
Method:                 Least Squares   F-statistic:                     57.46
Date:                Sun, 24 May 2020   Prob (F-statistic):           2.70e-20
Time:                        23:28:06   Log-Likelihood:                -482.88
No. Observations:                  88   AIC:                             973.8
Df Residuals:                      84   BIC:                             983.7
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -21.7703     29.475     -0.739      0.4

C5.iii The level model has a slightly higher R-squared, making it preferable

In [21]:
#Exercise C6
vote1 = pd.read_stata("stata/VOTE1.DTA")
vote1["expendAexpendB"] = vote1.expendA * vote1.expendB

y = vote1.voteA
X = sm.add_constant(vote1[["prtystrA", "expendA", "expendB", "expendAexpendB"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  voteA   R-squared:                       0.571
Model:                            OLS   Adj. R-squared:                  0.561
Method:                 Least Squares   F-statistic:                     55.86
Date:                Sun, 24 May 2020   Prob (F-statistic):           6.77e-30
Time:                        23:28:06   Log-Likelihood:                -659.75
No. Observations:                 173   AIC:                             1329.
Df Residuals:                     168   BIC:                             1345.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const             32.1174      4.591      6.

In [22]:
vote1.expendA.mean()

310.61090087890625

In [23]:
(model.params[3] + (model.params[4] * 300)) * 100

-3.37126929986698

In [24]:
(model.params[2] + (model.params[4] * 100)) * 100

3.7617990731507858

In [25]:
y = vote1.voteA
X = sm.add_constant(vote1[["prtystrA", "expendA", "expendB", "shareA"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  voteA   R-squared:                       0.868
Model:                            OLS   Adj. R-squared:                  0.865
Method:                 Least Squares   F-statistic:                     276.5
Date:                Sun, 24 May 2020   Prob (F-statistic):           9.03e-73
Time:                        23:28:06   Log-Likelihood:                -557.66
No. Observations:                 173   AIC:                             1125.
Df Residuals:                     168   BIC:                             1141.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         18.1954      2.568      7.086      0.0

In [26]:
-model.params[4] * (300 / (300 ** 2))

-0.0016479811989954254

C6.i It is not at all obvious what the sign of the interaction should be.

C6.ii The interaction is not statistically significant.

C6.iii The shift is about 3% of the vote. Given that the mean of Vote A is 50, this does mean the difference in the election.

C6.iv This increases the vote share by about 3.7. Slightly higher than expenditure on vote B but consistent with the idea that spending for both candidates are equal.

C6.v This does not have a lot of intuitive sense since the share is measuring the difference between the two candidates which should be captured by the expenditure values.

C6.vi The change is 0.0016. It implies that spending will reduce the vote share by 0.1% by adding expenditure

In [27]:
#Exercise C7
attend = pd.read_stata("stata/attend.dta")
attend["priGPAsq"] = attend.priGPA ** 2
attend["ACTsq"] = attend.ACT.astype("int") ** 2
attend["priGPAatndrte"] = attend.priGPA * attend.atndrte.astype("int")

y = attend.stndfnl
X = sm.add_constant(attend[["atndrte", "priGPA", "ACT", "priGPAsq", "ACTsq", "priGPAatndrte"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                stndfnl   R-squared:                       0.229
Model:                            OLS   Adj. R-squared:                  0.222
Method:                 Least Squares   F-statistic:                     33.33
Date:                Sun, 24 May 2020   Prob (F-statistic):           2.93e-35
Time:                        23:28:06   Log-Likelihood:                -868.72
No. Observations:                 680   AIC:                             1751.
Df Residuals:                     673   BIC:                             1783.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const             2.1164      1.357      1.560

In [28]:
print(model.params[2] + (2 * model.params[4] * 2.59) + (model.params[6] * .82))

# Compare this to text's estimates
print(-1.63 + (2 * 0.296 * 2.59) + (0.0056 * .82))

-0.13263760851809578
-0.09212800000000014


In [29]:
attend["new_b4"] = (attend.priGPA.astype("int") - 2.59) ** 2
attend["new_b6"] = attend.priGPA.astype("int") * (attend.atndrte.astype("int") - 82)

X = sm.add_constant(attend[["atndrte", "priGPA", "ACT", "new_b4", "ACTsq", "new_b6"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                stndfnl   R-squared:                       0.216
Model:                            OLS   Adj. R-squared:                  0.209
Method:                 Least Squares   F-statistic:                     30.85
Date:                Sun, 24 May 2020   Prob (F-statistic):           8.46e-33
Time:                        23:28:06   Log-Likelihood:                -874.57
No. Observations:                 680   AIC:                             1763.
Df Residuals:                     673   BIC:                             1795.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0091      1.319      0.007      0.9

C7.i The estimates appear different from the text but the calculated value from my estimation is -0.13

In [30]:
#Exercise C8
hprice1 = pd.read_stata("stata/hprice1.dta")

y = hprice1.price
X = sm.add_constant(hprice1[["lotsize", "sqrft", "bdrms"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.672
Model:                            OLS   Adj. R-squared:                  0.661
Method:                 Least Squares   F-statistic:                     57.46
Date:                Sun, 24 May 2020   Prob (F-statistic):           2.70e-20
Time:                        23:28:06   Log-Likelihood:                -482.88
No. Observations:                  88   AIC:                             973.8
Df Residuals:                      84   BIC:                             983.7
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -21.7703     29.475     -0.739      0.4

In [31]:
np.sqrt(model.mse_resid)

59.8334797507195

In [32]:
model.ssr / model.df_resid

3580.04529907976

In [33]:
X_p = pd.DataFrame({"const" : [1.0], "lotsize" : [10000], "sqrft": [2300], "bdrms": [4]})
print(model.predict(exog=X_p))

0    336.706671
dtype: float64


In [34]:
hprice1["adjLotsize"] = hprice1.lotsize - 10000
hprice1["adjSqrft"] = hprice1.sqrft - 2300
hprice1["adjBdrms"] = hprice1.bdrms - 4

y = hprice1.price
X = sm.add_constant(hprice1[["adjLotsize", "adjSqrft", "adjBdrms"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.672
Model:                            OLS   Adj. R-squared:                  0.661
Method:                 Least Squares   F-statistic:                     57.46
Date:                Sun, 24 May 2020   Prob (F-statistic):           2.70e-20
Time:                        23:28:06   Log-Likelihood:                -482.88
No. Observations:                  88   AIC:                             973.8
Df Residuals:                      84   BIC:                             983.7
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        336.7067      7.374     45.658      0.0

In [35]:
# By equation 6.36
se_e0 = np.sqrt((model.bse[0] ** 2) + model.mse_resid)

CI = (336.706671 - (1.96 * se_e0), 336.706671 + (1.96 * se_e0))
print(CI, "Width: ", CI[1] - CI[0])

(218.5456841629536, 454.86765783704635) Width:  236.32197367409276


C8.i Standard error is 59.83, Predicted price is 336.70

C8.ii Confidence interval is (322.042, 351.372)

C8.iii Width is ~236 which is very wide (which would be understandable since we haven't used much to explain the house price)

In [36]:
# Exercise 9
nbasal = pd.read_stata("stata/nbasal.dta")

y = nbasal.points
X = sm.add_constant(nbasal[["exper", "expersq", "age", "coll"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 points   R-squared:                       0.141
Model:                            OLS   Adj. R-squared:                  0.128
Method:                 Least Squares   F-statistic:                     10.85
Date:                Sun, 24 May 2020   Prob (F-statistic):           3.69e-08
Time:                        23:28:06   Log-Likelihood:                -838.21
No. Observations:                 269   AIC:                             1686.
Df Residuals:                     264   BIC:                             1704.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         35.2183      6.987      5.041      0.0

In [37]:
abs(model.params[1])/abs(2 * model.params[2])

15.342896101914247

In [38]:
X = sm.add_constant(nbasal[["exper", "expersq", "age", "agesq", "coll"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 points   R-squared:                       0.145
Model:                            OLS   Adj. R-squared:                  0.129
Method:                 Least Squares   F-statistic:                     8.925
Date:                Sun, 24 May 2020   Prob (F-statistic):           7.61e-08
Time:                        23:28:06   Log-Likelihood:                -837.61
No. Observations:                 269   AIC:                             1687.
Df Residuals:                     263   BIC:                             1709.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         73.5903     35.933      2.048      0.0

In [39]:
y = nbasal.lwage
X = sm.add_constant(nbasal[["points", "exper", "expersq", "age", "coll"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.488
Model:                            OLS   Adj. R-squared:                  0.478
Method:                 Least Squares   F-statistic:                     50.10
Date:                Sun, 24 May 2020   Prob (F-statistic):           2.43e-36
Time:                        23:28:06   Log-Likelihood:                -257.23
No. Observations:                 269   AIC:                             526.5
Df Residuals:                     263   BIC:                             548.0
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.7790      0.845      8.019      0.0

In [40]:
model.f_test("(age = 0), (coll = 0)")

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[1.18907125]]), p=0.3061355102306365, df_denom=263, df_num=2>

C9.ii Experience begins to have a negative effect on points at 15.34. Intuitively, this does not make sense since this is likely to be more related to age than anything

C9.iii If NBA players can be drafted before college then this means that the college players who remain are not 'first picks' and so staying later will select players with less ability

C9.iv Age is significant but the quadratic term is not.

C9.vi age and college experience are not jointly significance. If they have any seperate effects once productivity and seniority are accounted for, we do not have any evidence for it

In [41]:
# Exercise 10
bwght2 = pd.read_stata("stata/bwght2.dta")
bwght_sample = bwght2[["lbwght", "npvis", "npvissq"]].dropna()

y = bwght_sample.lbwght
X = sm.add_constant(bwght_sample[["npvis", "npvissq"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 lbwght   R-squared:                       0.021
Model:                            OLS   Adj. R-squared:                  0.020
Method:                 Least Squares   F-statistic:                     19.12
Date:                Sun, 24 May 2020   Prob (F-statistic):           6.10e-09
Time:                        23:28:07   Log-Likelihood:                 310.56
No. Observations:                1764   AIC:                            -615.1
Df Residuals:                    1761   BIC:                            -598.7
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.9579      0.027    291.364      0.0

In [42]:
abs(model.params[1] / (2 * model.params[2]))

22.059998742704906

In [43]:
bwght_sample[bwght_sample["npvis"] > 22].shape[0]

19

In [44]:
bwght_sample = bwght2[["lbwght", "npvis", "npvissq", "mage", "magesq"]].dropna()
X = sm.add_constant(bwght_sample[["npvis", "npvissq", "mage", "magesq"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
lbweight_fit = model.fittedvalues
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 lbwght   R-squared:                       0.026
Model:                            OLS   Adj. R-squared:                  0.023
Method:                 Least Squares   F-statistic:                     11.56
Date:                Sun, 24 May 2020   Prob (F-statistic):           2.87e-09
Time:                        23:28:07   Log-Likelihood:                 314.51
No. Observations:                1764   AIC:                            -619.0
Df Residuals:                    1759   BIC:                            -591.6
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.5837      0.137     55.333      0.0

In [45]:
abs(model.params[3] / (2 * model.params[4]))

30.82546325949819

In [46]:
bwght_sample[bwght_sample["mage"] > abs(model.params[3] / (2 * model.params[4]))].shape[0]

720

In [47]:
bwght_sample = bwght2[["bwght", "npvis", "npvissq", "mage", "magesq"]].dropna()
y = bwght_sample.bwght
X = sm.add_constant(bwght_sample[["npvis", "npvissq", "mage", "magesq"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
bweight_fit = model.fittedvalues
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  bwght   R-squared:                       0.019
Model:                            OLS   Adj. R-squared:                  0.017
Method:                 Least Squares   F-statistic:                     8.590
Date:                Sun, 24 May 2020   Prob (F-statistic):           7.29e-07
Time:                        23:28:07   Log-Likelihood:                -13711.
No. Observations:                1764   AIC:                         2.743e+04
Df Residuals:                    1759   BIC:                         2.746e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1860.3805    388.977      4.783      0.0

C10.i The quadratic term is statistically significant but quite small

C10.ii 19 women had greater tan 22 prenatalvisits.

C10.iii If there is a problem, it is more likely that women will make more visits

C10.iv The age is maximized at approximately 30 (a little over), 720 women (about 40% of the sample) are older than 30

C10.v The log form only explains about 2% of the variation

C10.vi We cannot directly compare r-squareds, but I cannot find the method recommended in the text (section 6.4).

In [48]:
# Exercise 11
apple = pd.read_stata("stata/APPLE.DTA")

y = apple.ecolbs
X = sm.add_constant(apple[["ecoprc", "regprc"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 ecolbs   R-squared:                       0.036
Model:                            OLS   Adj. R-squared:                  0.033
Method:                 Least Squares   F-statistic:                     12.41
Date:                Sun, 24 May 2020   Prob (F-statistic):           5.11e-06
Time:                        23:28:07   Log-Likelihood:                -1535.3
No. Observations:                 660   AIC:                             3077.
Df Residuals:                     657   BIC:                             3090.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.9653      0.380      5.171      0.0

In [49]:
(model.fittedvalues.min(), model.fittedvalues.max())

(0.8551603344584171, 2.087215047435918)

In [50]:
apple[apple.ecolbs == 0].shape[0] / apple.shape[0]

0.37575757575757573

In [51]:
X = sm.add_constant(apple[["ecoprc", "regprc", "faminc", "hhsize", "educ", "age"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 ecolbs   R-squared:                       0.040
Model:                            OLS   Adj. R-squared:                  0.031
Method:                 Least Squares   F-statistic:                     4.561
Date:                Sun, 24 May 2020   Prob (F-statistic):           0.000152
Time:                        23:28:07   Log-Likelihood:                -1534.0
No. Observations:                 660   AIC:                             3082.
Df Residuals:                     653   BIC:                             3113.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.0568      0.893      1.184      0.2

In [52]:
model.f_test("(faminc = 0), (hhsize = 0), (educ = 0), (age = 0)")

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[0.64785687]]), p=0.6285503171166624, df_denom=653, df_num=4>

In [53]:
X = sm.add_constant(apple[["ecoprc"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 ecolbs   R-squared:                       0.010
Model:                            OLS   Adj. R-squared:                  0.008
Method:                 Least Squares   F-statistic:                     6.501
Date:                Sun, 24 May 2020   Prob (F-statistic):             0.0110
Time:                        23:28:07   Log-Likelihood:                -1544.3
No. Observations:                 660   AIC:                             3093.
Df Residuals:                     658   BIC:                             3102.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.3881      0.372      6.426      0.0

In [54]:
X = sm.add_constant(apple[["regprc"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 ecolbs   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                   0.04937
Date:                Sun, 24 May 2020   Prob (F-statistic):              0.824
Time:                        23:28:07   Log-Likelihood:                -1547.5
No. Observations:                 660   AIC:                             3099.
Df Residuals:                     658   BIC:                             3108.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.3950      0.369      3.782      0.0

In [55]:
apple.ecolbs.corr(apple.regprc)

0.00866310567899933

C11.i Consumption of eco apples decreses with price and increases when the price of a susbtitute (regular) rises

C11.ii Both are statistically significant with p-values of 0.000

C11.iii Fitted values range from 0.85 to 2.05. 38% of the sample does not consume any

C11.iv The price variables explain less than 10% of the variation

C11.v The joint significance of the added values has a p-value of 0.6 and so we do not have evidence to suggest they significantly affect consumption of eco apples

C11.vi They are much smaller but have the same signs. regprc loses its significance, likely due to the fact that the two variables do not appear to be correlated

In [56]:
# Exercise 12
k401subs = pd.read_stata("stata/401ksubs.dta")
k401subs_sub = k401subs[k401subs.fsize == 1]

k401subs_sub.age.min()

25

In [57]:
k401subs_sub[k401subs_sub.age == 25].shape[0]

99

In [58]:
y = k401subs_sub.nettfa
X = sm.add_constant(k401subs_sub[["inc", "age", "agesq"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 nettfa   R-squared:                       0.123
Model:                            OLS   Adj. R-squared:                  0.122
Method:                 Least Squares   F-statistic:                     93.99
Date:                Sun, 24 May 2020   Prob (F-statistic):           6.30e-57
Time:                        23:28:07   Log-Likelihood:                -10520.
No. Observations:                2017   AIC:                         2.105e+04
Df Residuals:                    2013   BIC:                         2.107e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.2042     15.281     -0.079      0.9

In [59]:
k401subs_sub["adjusted_b3"] = (k401subs_sub.age.astype("int") - 25) ** 2
X = sm.add_constant(k401subs_sub[["inc", "age", "adjusted_b3"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 nettfa   R-squared:                       0.123
Model:                            OLS   Adj. R-squared:                  0.122
Method:                 Least Squares   F-statistic:                     93.99
Date:                Sun, 24 May 2020   Prob (F-statistic):           6.30e-57
Time:                        23:28:07   Log-Likelihood:                -10520.
No. Observations:                2017   AIC:                         2.105e+04
Df Residuals:                    2013   BIC:                         2.107e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const         -17.1807      9.973     -1.723      

/home/systemchalk/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [60]:
k401subs_sub["adjusted_b3"] = (k401subs_sub.age.astype("int") - 25) ** 2
X = sm.add_constant(k401subs_sub[["inc",  "adjusted_b3"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                 nettfa   R-squared:                       0.123
Model:                            OLS   Adj. R-squared:                  0.122
Method:                 Least Squares   F-statistic:                     141.0
Date:                Sun, 24 May 2020   Prob (F-statistic):           4.73e-58
Time:                        23:28:07   Log-Likelihood:                -10520.
No. Observations:                2017   AIC:                         2.105e+04
Df Residuals:                    2014   BIC:                         2.106e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const         -18.4881      2.178     -8.490      

/home/systemchalk/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


C12.i Youngest age is 25 with 99 at that age

C12.ii Increase in age holding all other factors, including age squared, fixed. This isn't particularly helpful on its own.

C12.iii Results reported above. The negative coefficient for age isn't as much of a concern because age squared is positive.

C12.iv Results above

C12.v The model doesn't fit entirely that much better in terms of goodness of fit.

In [61]:
# Exercise 13
meap00 = pd.read_stata("stata/meap00_01.dta")

y = meap00.math4
X = sm.add_constant(meap00[["lexppp", "lenroll", "lunch"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  math4   R-squared:                       0.373
Model:                            OLS   Adj. R-squared:                  0.372
Method:                 Least Squares   F-statistic:                     334.6
Date:                Sun, 24 May 2020   Prob (F-statistic):          1.83e-170
Time:                        23:28:07   Log-Likelihood:                -7014.6
No. Observations:                1692   AIC:                         1.404e+04
Df Residuals:                    1688   BIC:                         1.406e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         91.9325     19.962      4.605      0.0

In [62]:
(model.fittedvalues.min(), model.fittedvalues.max())

(42.41416403600728, 92.67098076453601)

In [63]:
(meap00.math4.min(), meap00.math4.max())

(0.0, 100.0)

In [64]:
meap00["resid"] = model.resid

In [65]:
meap00["bcode"][meap00.resid == model.resid.max()]

1175    1141
Name: bcode, dtype: int16

In [66]:
meap00["lexpppsq"] = meap00.lexppp ** 2
meap00["lenrollsq"] = meap00.lenroll ** 2
meap00["lunchsq"] = meap00.lunch ** 2

X = sm.add_constant(meap00[["lexppp", "lenroll", "lunch", "lexpppsq", "lenrollsq", "lunchsq"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                  math4   R-squared:                       0.373
Model:                            OLS   Adj. R-squared:                  0.371
Method:                 Least Squares   F-statistic:                     167.4
Date:                Sun, 24 May 2020   Prob (F-statistic):          4.21e-167
Time:                        23:28:08   Log-Likelihood:                -7013.9
No. Observations:                1692   AIC:                         1.404e+04
Df Residuals:                    1685   BIC:                         1.408e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         79.1976    334.604      0.237      0.8

In [67]:
model.f_test("(lexpppsq = 0), (lenrollsq = 0), (lunchsq = 0)")

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[0.51802367]]), p=0.6699065564363909, df_denom=1.68e+03, df_num=3>

In [68]:
meap00["lexpppstd"] = meap00.lexppp / meap00.lexppp.std()
meap00["lenrollstd"] = meap00.lenroll / meap00.lenroll.std()
meap00["lunchstd"] = meap00.lunch / meap00.lunch.std()
meap00["math4std"] = meap00.math4 / meap00.math4.std()

y = meap00.math4std
X = sm.add_constant(meap00[["lexpppstd", "lenrollstd", "lunchstd"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:               math4std   R-squared:                       0.373
Model:                            OLS   Adj. R-squared:                  0.372
Method:                 Least Squares   F-statistic:                     334.6
Date:                Sun, 24 May 2020   Prob (F-statistic):          1.83e-170
Time:                        23:28:08   Log-Likelihood:                -2005.6
No. Observations:                1692   AIC:                             4019.
Df Residuals:                    1688   BIC:                             4041.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          4.7618      1.034      4.605      0.0

C13.i All but lexppp

C13.ii (42.4141640360071, 92.670980764536) vs (0, 100)

C13.iii Building code 1141 has the largest positive residual. This school had an uncommonly strong math result

C13.iv The f-test fails to provide evidence that any of the quadratic terms are significant. There is little reason to leave them in

C13.v The largest change in terms of standard deviation is brought on by the lunch variable (negative effect)

In [69]:
# Exercise 14
benefits = pd.read_stata("stata/benefits.dta")

y = benefits.lavgsal
X = sm.add_constant(benefits[["bs"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                lavgsal   R-squared:                       0.005
Model:                            OLS   Adj. R-squared:                  0.004
Method:                 Least Squares   F-statistic:                     9.182
Date:                Sun, 24 May 2020   Prob (F-statistic):            0.00248
Time:                        23:28:08   Log-Likelihood:                 75.732
No. Observations:                1848   AIC:                            -147.5
Df Residuals:                    1846   BIC:                            -136.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         10.6476      0.057    186.018      0.0

In [70]:
model.t_test("bs = -1")

<class 'statsmodels.stats.contrast.ContrastResults'>
                             Test for Constraints                             
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
c0            -0.5035      0.166      2.988      0.003      -0.829      -0.178

In [71]:
benefits["lbs"] = np.log(benefits.bs)
print(benefits.lbs.min(), benefits.lbs.max(), benefits.lbs.std())

X = sm.add_constant(benefits[["lbs"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

-2.33083176612854 -0.416291207075119 0.0968237817287445
                            OLS Regression Results                            
Dep. Variable:                lavgsal   R-squared:                       0.004
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     7.241
Date:                Sun, 24 May 2020   Prob (F-statistic):            0.00719
Time:                        23:28:08   Log-Likelihood:                 74.765
No. Observations:                1848   AIC:                            -145.5
Df Residuals:                    1846   BIC:                            -134.5
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------

In [72]:
X = sm.add_constant(benefits[["bs", "lenroll", "lstaff", "lunch"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                lavgsal   R-squared:                       0.483
Model:                            OLS   Adj. R-squared:                  0.481
Method:                 Least Squares   F-statistic:                     429.8
Date:                Sun, 24 May 2020   Prob (F-statistic):          8.56e-262
Time:                        23:28:08   Log-Likelihood:                 680.03
No. Observations:                1848   AIC:                            -1350.
Df Residuals:                    1843   BIC:                            -1322.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         13.7236      0.112    122.413      0.0

In [73]:
benefits["lunchsq"] = benefits.lunch ** 2

X = sm.add_constant(benefits[["bs", "lenroll", "lstaff", "lunch", "lunchsq"]])
model = sm.OLS(y, X).fit()
model_summary = model.summary()
print(model_summary)

                            OLS Regression Results                            
Dep. Variable:                lavgsal   R-squared:                       0.491
Model:                            OLS   Adj. R-squared:                  0.490
Method:                 Least Squares   F-statistic:                     355.6
Date:                Sun, 24 May 2020   Prob (F-statistic):          4.33e-267
Time:                        23:28:08   Log-Likelihood:                 695.39
No. Observations:                1848   AIC:                            -1379.
Df Residuals:                    1842   BIC:                            -1346.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         13.8334      0.113    122.471      0.0

In [74]:
abs(model.params[4] / (2 * model.params[5]))

56.63195040623062

In [75]:
benefits[benefits.lunch > abs(model.params[4] / (2 * model.params[5]))].shape[0]

418

C14.i We can reject the null at all levels. P-values are above but the one sided test must be divided by two

C14.ii Reported above

C14.iii It does not

C14.iv Coefficient ceases to be different from zero

C14.v More staff hired when labour is cheap or, alternatively, HR budget must be spread thin

C14.vi The quadratic is statistically significant with a turning point at 56.63. There are 418 values above this point

C14.vii Teacher salaries are lower with school poverty rates up to over half of the school being in poverty. IN this case, lunch=100 would be the ideal school